### PR TITLE
[release-4.8] Bug 2066365: (Topology) Performance improvement by reducing rerenderings and deep-copy toJSON() calls

### PR DESCRIPTION
**Pull request by jerolimov**
Manual cherry-pick of #11184 after automatical cherry-pick #11209 doesn't compile and the tests don't pass because there was no react testing library in 4.8.

In the 2nd commit, I fixed the import ref issue and downgraded the unit tests from `@testing-library/react` to `react-dom/test-utils`

Will add some screenshots about the performance difference soon (tomorrow)...

### CODE REVIEW

**Comment by jerolimov** [comment]
@invincibleJai fixed an issue in release-4.9 branch, see #11256, we need to analyze and maybe include this into this PR before merging.

/hold

**Comment by invincibleJai** [comment]
@karthikjeeyar @rottencandy can one of you verify this I did verify with clusterbot to see the issue with git import flow which we observed in 4.9 but I don't see it here

image

**Review by rottencandy — APPROVED** [review]
Tested on cluster, git page does not crash after load. And topology is functional without lag even when working with 100 deployments.

https://user-images.githubusercontent.com/20013884/161989942-31bae6a9-ab94-4c1f-8ede-4bf4b9f25490.mp4

**Comment by jerolimov** [comment]
TODO

/unhold

**Pull request merged**
